File: ios/Packages/SimplicityServices/Sources/SimplicityServices/LearningService.swift
import Foundation
import Mockable
import SimplicityApi

/// Everything the Learn tab needs from the API.
///
/// Nothing here computes progress. Status, completion and marking are all the server's answers,
/// which is what keeps the app and the web agreeing about the same clinician.
@Mockable
public protocol LearningService: AnyObject, Sendable {
    func assignedModules(orgId: UUID) async throws -> [AssignedModuleResponse]
    func module(orgId: UUID, moduleId: UUID) async throws -> LearnerModuleResponse
    func quiz(orgId: UUID, moduleId: UUID) async throws -> QuizResponse

    @discardableResult
    func completeSection(orgId: UUID, sectionId: UUID) async throws -> LearnerModuleResponse

    func submitAttempt(
        orgId: UUID,
        moduleId: UUID,
        answers: [AnswerInput]
    ) async throws -> AttemptResultResponse

    func playback(orgId: UUID, assetId: UUID) async throws -> PlaybackResponse
}

public extension AssignedModuleResponse {

    /// Anything not finished, including a module that was finished before its content changed
    /// substantively and has come back around.
    var isOutstanding: Bool {
        status != .completed
    }
}

File: android/services/src/main/kotlin/io/simplicity/training/services/LearningService.kt
package io.simplicity.training.services

import io.simplicity.training.api.apis.LearningApi
import io.simplicity.training.api.models.AnswerInput
import io.simplicity.training.api.models.AssignedModuleResponse
import io.simplicity.training.api.models.AttemptResultResponse
import io.simplicity.training.api.models.LearnerModuleResponse
import io.simplicity.training.api.models.PlaybackResponse
import io.simplicity.training.api.models.QuizResponse
import io.simplicity.training.api.models.SubmitAttemptRequest
import java.util.UUID

/**
 * Everything the Learn screens need from the API.
 *
 * Nothing here computes progress. Status, completion and marking are all the server's answers,
 * which is what keeps the app and the web agreeing about the same clinician.
 */
interface LearningService {
    suspend fun assignedModules(orgId: UUID): List<AssignedModuleResponse>
    suspend fun module(orgId: UUID, moduleId: UUID): LearnerModuleResponse
    suspend fun quiz(orgId: UUID, moduleId: UUID): QuizResponse
    suspend fun completeSection(orgId: UUID, sectionId: UUID): LearnerModuleResponse
    suspend fun submitAttempt(orgId: UUID, moduleId: UUID, answers: List<AnswerInput>): AttemptResultResponse
    suspend fun playback(orgId: UUID, assetId: UUID): PlaybackResponse
}

class LearningServiceImpl(private val api: LearningApi) : LearningService {

    override suspend fun assignedModules(orgId: UUID) = api.listAssignedModules(orgId).unwrap()

    override suspend fun module(orgId: UUID, moduleId: UUID) = api.readModule(orgId, moduleId).unwrap()

    override suspend fun quiz(orgId: UUID, moduleId: UUID) = api.getQuiz(orgId, moduleId).unwrap()

    override suspend fun completeSection(orgId: UUID, sectionId: UUID) =
        api.completeSection(orgId, sectionId).unwrap()

    override suspend fun submitAttempt(orgId: UUID, moduleId: UUID, answers: List<AnswerInput>) =
        api.submitQuizAttempt(orgId, moduleId, SubmitAttemptRequest(answers)).unwrap()

    override suspend fun playback(orgId: UUID, assetId: UUID) = api.getPlaybackUrl(orgId, assetId).unwrap()
}

/** Anything not finished, including a module that changed substantively and has come back around. */
val AssignedModuleResponse.isOutstanding: Boolean
    get() = status != AssignedModuleResponse.Status.COMPLETED
